src/libcore/iter.rs

```diff
@@ -1791,6 +1797,16 @@ impl<T, I> Iterator for Peekable<T, I> where I: Iterator<Item=T> {
 }
 
 #[stable]
+impl<T, I> ExactSizeIterator for Peekable<T, I> where I: ExactSizeIterator<Item = T> {
+    #[inline]
+    fn len(&self) -> usize {
+        // This is guarenteed to not overflow because `len()` must have been able to return a valid
+        // value before we peeked.
+        self.iter.len() + if self.peeked.is_some() { 1 } else { 0 }
+    }
+}
```

Comment:
We can because, given an underlying iterator that implements `ExactSizeIterator`, either:
  1. We haven't peeked: `self.peeked` is `None` and we forward the result from `self.iter.size_hint()` which must meet the `ExactSizeIterator` spec.
  2. We have peeked: To have peeked, we must have called `next()` on `self.iter` so `self.iter`'s `size_hint()` must be at most `(lo, Some(hi)` where `lo == hi <= uint::MAX-1` because it must have met the `ExactSizeIterator` spec both before and after we peeked. Therefore adding 1 cannot overflow.